test: Support booting Flynn in Flynn!

lmars opened the pull request:
This adds support for booting Flynn clusters inside Flynn clusters :smile:.

Summary:

* the `test/cluster2` package (which will eventually replace `test/cluster`) boots Flynn clusters by creating a flynn-host app, scaling it up, then running the bootstrapper once the hosts have started
* flynn-test now has a `bootCluster` function which boots a cluster and returns an object which can be used to run commands against the cluster
* flynn-test now runs in a Flynn job, which means it can happily change `/etc/hosts` at will, and also use discoverd for DNS without [hacking /etc/resolv.conf](https://github.com/flynn/flynn/blob/v20161114.0p1/script/run-integration-tests#L92-L94)
* flynn-host needs extra Linux capabilities (`CAP_SYS_ADMIN` and `CAP_NET_ADMIN`), needs to access `/dev/zfs` and `/dev/zd*`, and needs write access to `cgroupfs`, so I've added support for customising these through `ct.ProcessType` and `host.ContainerConfig`
* I've added a xenial based `host` image which has the `zfsutils-linux` package and a custom udev rule to create ZFS device nodes in the host's mount namespace
* flynn-host jobs have `/var/lib/flynn` mounted from the host and create volumes in `/var/lib/flynn/JOB_ID/volumes` (`zfs` doesn't like the sparse vdev file being in the container's filesystem, it needs to be available at the same path both inside and outside the container)
* some disruptive tests have been updated to boot their own clusters

This is a first step in refactoring CI (see #2970). I have some grand plans, but one step at a time :wink:.

Lets make CI great again!

titanous commented on host/libcontainer_backend.go:
```diff
@@ -448,7 +453,7 @@ func (l *LibcontainerBackend) Run(job *host.Job, runConfig *RunConfig, rateLimit
 			Path: filepath.Join("/flynn", job.Partition, job.ID),
 			Resources: &configs.Resources{
-				AllowedDevices: configs.DefaultAllowedDevices,
+				AllowedDevices: append(configs.DefaultAllowedDevices, job.Config.AllowedDevices...),
```
In order to allow restricting jobs further, it probably makes sense to use _only_ the provided devices and capabilities if specified so that the defaults can be overridden?

titanous commented on test/cluster2/cluster.go:
```diff
@@ -0,0 +1,340 @@
+		Processes: map[string]int{"host": 0},
+	}); err != nil {
+		return err
+	}
```
Doesn't the `DeleteApp` below do this?

lmars commented:
I have updated some more tests to boot their own clusters (1bae5e5) including ones which add hosts to existing clusters (f1a0227).

Booting multi-node clusters inside multi-node clusters doesn't seem to work (the Flannel overlay network packets don't seem to make it to the right place), but multi-node within single-node works fine. Given tests can boot multi-node clusters if needed, I have changed the default CI cluster to be single-node (see c305723).

The multi-node tests will fail here, but I have tested in custom CI so should pass once we merge / restart CI.

@titanous I have addressed your comments too.

Ready for a review.

lmars commented:
I changed the backups to use unique domain names in 9a18478 so we can run the backup tests in parallel but this has broken the tests on master which expect the domain to be `1.localflynn.com` for all the backups (see [here](https://github.com/flynn/flynn/blob/master/test/test_zz_backup.go#L137)).

The sooner we can merge this the better (the only failing tests are the ones which try and boot multi-node clusters, which will pass once we restart CI).

If there are any blockers for this though I'll revert the backups in CI.
